File: 200. Number of Islands/200. Number of Islands.java
class Solution {

  int m, n;

  public int numIslands(char[][] grid) {
    m = grid.length;
    n = grid[0].length;

    int numIslands = 0;

    for (int i = 0; i < m; i++) {
      for (int j = 0; j < n; j++) {
        if (grid[i][j] == '1') {
          dfs(i, j, grid);
          numIslands++;
        }
      }
    }

    return numIslands;
  }

  private void dfs(int i, int j, char[][] grid) {
    if (i < 0 || i >= m || j < 0 || j >= n || grid[i][j] != '1') {
      return;
    }

    grid[i][j] = '0';

    dfs(i + 1, j, grid);
    dfs(i - 1, j, grid);
    dfs(i, j + 1, grid);
    dfs(i, j - 1, grid);
  }
}

File: 200. Number of Islands.py
class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        ROWS, COLS = len(grid), len(grid[0])
        num_islands = 0

        for r in range(ROWS):
            for c in range(COLS):
                if grid[r][c] == "1":
                    queue = deque([(r, c)])
                    grid[r][c] = "0"
                    while queue:
                        cr, cc = queue.popleft()
                        for dr, dc in [(1, 0), (-1, 0), (0, 1), (0, -1)]:
                            nr, nc = cr + dr, cc + dc
                            if 0 <= nr < ROWS and 0 <= nc < COLS and grid[nr][nc] == "1":
                                grid[nr][nc] = "0"
                                queue.append((nr, nc))
                    num_islands += 1

        return num_islands
